File: find_largest.py
import os, time, pathlib, hashlib, random, tkinter
from tkinter import filedialog

# find largest file
class find_largest():
    def __init__(self):
        self.step_size = 100
        self.files_size = []
        self.count = 0
        self.files = []
        self.print_time = 5
        self.fps = 0
        self.paths = {}
        self.log_filename = "output.log"
        self.dirs_count = 0
        self.stop_count = False
        self.dir_first_run = True
        self.search_by_content = False
        self.total_dir_count = 0
        self.search_results = []
        self.file_types = ['.ods', '.mhtml', '.ml', '.otx', '.xlam', '.ptm', '.xlsb', '.ps', '.xps', '.xlw',
            '.xlsm', '.dot', '.mht', '.dif', '.xlt', '.doc', '.xltm', '.txt', '.dp', '.psm', '.csv', '.odt',
            '.docx', '.xml', '.pam', '.htm', '.dotm', '.xls', '.xlsx', '.ptx', '.docm', '.rtf', '.pdf', '.otm',
            '.htm, .html', '.xltx', '.prn', '.wps', '.df', '.psx', '.ps', '.html', '.slk', '.xla', '.dotx',
            '.tf', '.ot', '.dbf', '.log', '.py']
        
    def walk(self, args):
        """ Walk through files in file system, lists all files and sub files"""
        self.files_size = []
        self.count = 0
        self.start_time = time.time()
        name = args[0]
        do_dirs = args[1]
        do_subdirs, self.search_term = args[2], args[3].lower()
        self.walking_in = ["folders" if do_dirs == True else "files"][0]
        self.update_log("", new_log=True)
        self.print_log('[+] walking '+self.walking_in+' in ' + name)
        for path, dirs, files in os.walk(name):
            # process folders
            if do_dirs == True:
                # process the folder
                if do_subdirs == False:
                    if path not in self.paths:
                        sub_files = []
                        for f in files:
                            try:
                                p = os.path.join(path, f)
                                sub_files.append(os.path.getsize(p))
                                self.update_fps()

                                # search file
                                if self.search_term != b"":
                                    # search by content
                                    if self.search_by_content == True:
                                        extension = "." + p.split("\\")[-1].split(".")[-1]
                                        if extension in self.file_types:
                                            # search name
                                            if self.search_term in f.encode("utf-8"):
                                                self.search_results.append(p)
                                            else:
                                                # search contents
                                                searched_file = open(p, "rb").read().lower()
                                                if self.search_term in searched_file:
                                                    self.search_results.append(p)
                                            
                                            
                                    # search by name
                                    else:
                                        if self.search_term in f.encode("utf-8"):
                                            self.search_results.append(p)
                                            
                            except Exception as error:
                                self.print_log("[-] Error thrown when process file!", error)
                        self.paths[path] = sum(sub_files)

                # process sub folders
                elif do_subdirs == True:
                    # get size of sub files
                    sub_folder_size = 0
                    for sub_path, sub_dir, sub_files in os.walk(path):
                        for sub_file in sub_files:
                            try:
                                p = os.path.join(sub_path, sub_file)
                                sub_folder_size += os.path.getsize(p)
                                self.update_fps()
                            except Exception as error:
                                self.print_log("[-] Error thrown when process file!", error)
                        if self.dir_first_run == True:
                            self.total_dir_count += len(sub_dir)

                    # add to dict
                    self.paths[path] = sub_folder_size
                    self.dirs_count += len(dirs)
                    self.dir_first_run = False
                    
            # process files
            else:
                for file in files:
                    try:
                        p = os.path.join(path, file)
                        size = os.path.getsize(p)
                        self.files_size.append([size, p])
                        self.update_fps()
                    except PermissionError:
                        self.print_log("[-] Failed to read file! Access is denied")
                    except Exception as error:
                        self.print_log("[-] Failed to read file!", error)
                    
        self.print_log("[+] Finished walking!")

        if do_dirs == True:
            temp_list = []
            for key in self.paths.keys():
                temp_list.append([self.paths[key], key])
            self.files = list(reversed(sorted(temp_list)))
        else:
            self.files = list(reversed(sorted(self.files_size)))
        self.write_csv()

    def update_fps(self):
        """ Calculate Files Per Second (FPS)"""
        self.count += 1
        self.fps += 1
        if self.dirs_count > 0 and self.stop_count == False:
            self.stop_count = True
            self.final_file_count = self.count
        if self.count % self.step_size == 0:
            if time.time()-self.start_time > self.print_time:
                self.start_time = time.time()
                self.cfps =  "{:,}".format(int(self.fps / self.print_time))
                self.fps = 0
                
                # update log
                if self.stop_count == True:
                    prct = round((self.dirs_count / self.total_dir_count)*100, 2)
                    dir_test = " files! fps="+self.cfps + " dirs="+str(self.dirs_count)+" prct="+str(prct)
                    self.print_log('[+] Walked', "{:,}".format(self.final_file_count) + dir_test)
                else:
                    self.print_log('[+] Walked', "{:,}".format(self.count), "files! fps="+self.cfps)

    def update_log(self, data, new_log=False, close=False):
        """ Create log file if it doesn't exist"""
        if new_log == True:
            self.print_log("[+] New log started!")
            open(self.log_filename, "w").close()
            self.log_handler = open(self.log_filename, "ab")
            
        elif os.path.exists(self.log_filename):
            try:
                self.log_handler.write(data.encode("utf-8") + b"\n")
            except Exception as error:
                print("[-] Failed to write to log!", error)
        elif os.path.exists(self.log_filename) == False:
            open(self.log_filename, "w").close()

        if close == True:
            self.log_handler.close()

    def print_log(self, *text):
        """ Print text and write it to log"""
        txt = " ".join([str(i) for i in list(text)])
        print(txt)
        self.update_log(txt)

    def gb(self, num):
        """ Convert bytes to diffrent sizes of storage"""
        terms = [" bytes", "KB", "MB", "GB", "TB", "PB"]
        count = 0
        if type(num) == str:
            if num.isdigit()==True:
                num = float(num)
            else:
                return num
            
        while num > 1024:
            num = num / 1024
            count += 1

        if "." in str(num):
            return (str(num).split('.')[0]+"."+str(num).split('.')[1][:2]) + terms[count]
        else:
            return str(num) + terms[count]

    def first_100(self, step=100):
        """ print the first X number of results"""
        # folders and files
        self.print_log("\n", step, "largest "+self.walking_in+":\n"+("-"*30))
        count = 1
        for file in self.files[0:step]:
            self.print_log(str(count)+"|\t" + self.gb(file[0]) + "\t" + file[1].replace('\\', '/'))
            count += 1
        self.update_log("", new_log=False, close=True)

        # file search results
        if self.search_term != b"":
            print("\n Search Results for "+str(self.search_term)[1:]+":\n"+("-"*30))
            if len(self.search_results) == 0:
                print("No results found!")
            else:
                for i,result in enumerate(self.search_results):
                    print(str(i+1)+"|\t" + result.replace("\\", "/"))
        os.popen("pause")

    def write_csv(self):
        """ Write the results to a CSV file"""
        # paths
        if len(self.paths.keys()) > 0:
            print("[+] Writing Paths to CSV!")
            path_writer = open("paths.csv", "wb")
            for key in self.paths.keys():
                path_writer.write((str(self.paths[key])+","+key+"\n").encode("utf-8"))
            path_writer.close()

        # files
        if len(self.files) > 0:
            print("[+] Writing Files to CSV!")
            file_writer = open("files.csv", "wb")
            for item in self.files:
                file_writer.write((str(item[0])+","+item[1]+"\n").encode("utf-8"))
            file_writer.close()

    def get_input(self, finder):
        """ Get input from user via console"""
        # return sys drive if input empty
        user_dir = input('Dir Path: ')
        if user_dir == "":
            user_dir = os.getenv("SystemDrive") + "\\"
        else:
            # check if input is valid path
            if os.path.exists(user_dir) == False:
                self.print_log("Nov a valid directory, please enter a valid dir e.g. C:\\")
            while os.path.exists(user_dir) == False:
                user_dir = input('Invalid Directory\nDir Path: ')

        # files or folders
        self.step_size = 100
        user_files = input("Menu:\n1. Files\n2. Folders\n3. Folders and sub folders (very slow)" +\
                           "\n4. Search files by content (very slow)\n5. Search files by name (fast)\n> ")
        if user_files[0:6].lower() in ["folder", "2"]:
            return [user_dir, True, False, b""]
        elif user_files[0:6].lower() in ["sub", "subfolders", "3"]:
            return [user_dir, True, True, b""]
        elif user_files[0:6].lower() in ["s", "search", "4"]:
            self.step_size = 1
            self.search_by_content = True
            search = input("search term: ").encode("utf-8")
            return [user_dir, True, False, search]
        elif user_files[0:6].lower() in ["n", "search name", "5"]:
            search = input("search term: ").encode("utf-8")
            return [user_dir, True, False, search]
        else:
            return [user_dir, False, False, b""]

def finder_main():
    """ The main function for finder"""
    finder = find_largest()
    while True:
        finder.walk(finder.get_input(finder))
        finder.first_100(step=10)
        menu = input("Press enter to continue or type e to exit...")
        if menu == "e":
            break

# Sort files
class sort_files():
    def __init__(self):
        self.peper = b"JaredBot1234_"
        self.check_count = 200
        self.chunk_size = 65536

    def hash_file(self, path):
        """ Generates MD5 hash of file"""
        md5 = hashlib.md5()
        with open(path, 'rb') as f:
            while True:
                data = f.read(self.chunk_size)
                if not data:
                    break
                md5.update(data)
                
        return md5.hexdigest()

    def remove_duplicates(self, path):
        """ Deletes duplicate files in a directory"""
        count = 0
        unique = []
        for filename in os.listdir(path):
            full_path = os.path.join(path, filename)
            if os.path.isfile(full_path):
                filehash = self.hash_file(full_path)
                if filehash not in unique: 
                    unique.append(filehash)
                else: 
                    os.remove(full_path)
                    print("[+] removed duplicate " + filename)
            count += 1
            if count % self.check_count == 0:
                print("checking file " + str(filename))
                
    def rename_files(self, path):
        """ Renames files in a directory"""
        temp_name = "temp_" + str(int(random.random()*10000))
        for fname in [temp_name, ""]:
            files = os.listdir(path)
            count = 0
            for index, file in enumerate(files):
                extension = file.split(".")[-1]
                if extension in ["jpg", "png", "gif", "mp4"]:
                    current_fname = (fname + "%05d"%count).encode("utf-8")
                    hashed = hashlib.md5(self.peper + current_fname).hexdigest()
                    os.rename(os.path.join(path, file), os.path.join(path, hashed+"."+extension))
                    count += 1

                    if count % 200 == 0:
                        print("["+str(count)+"]renamed file " + file)

def remove_dupes_main():
    # path
    sort = sort_files()
    path = input("Dir Path: ")
    if path == "":
        path = pathlib.Path(__file__).parent.absolute()
    input("[+] Duplicate files in '"+str(path)+"' will be removed!\nPress enter to continue...")

    # settings
    itter = input("No. itterations: ")
    itter = ["1" if len(itter)==0 else itter][0]
    while itter.isdigit() == False:
        itter = input("Invalid Input!\nNo. itterations: ")

    print("Starting!")
    for i in range(1, int(itter)+1):
        sort.remove_duplicates(path)
        print("[+] Itteration "+str(i)+"/"+str(itter)+" Complete!")

    # rename files
    sort.rename_files(path)

class GUI():
    def __init__(self):
        self.row_counter = 0
        self.select_dir = False
        self.text_input = None

    def get_input_cb(self, win, tk, cb, args):
        result = tk.get(1.0, "end")
        self.text_input = result
        win.withdraw()
        self.top.withdraw()

        txt = (lambda x : "\n".join(x[:-1]) if x[-1]=="" else "\n".join(x))(self.text_input.split("\n"))
        txt = txt.encode("utf-8")
        cb.walk([args[0], args[1], args[2], txt])
        cb.first_100(step=10)

    def get_input(self, cb, args):
        win = tkinter.Tk()
        txt = tkinter.Text(win, height=10)
        txt.pack()

        command = lambda: self.get_input_cb(win, txt, cb, args)
        btn1 = tkinter.Button(win, height=1, width=10, text="Submit", command=command)
        btn1.pack()
    
    def button_callback(self, arg):
        finder = find_largest()

        if self.select_dir == True:
            print("Please Select a directory!")
            return
        
        self.select_dir = True
        user_dir = filedialog.askdirectory()
        
        if arg == 1:
            self.top.withdraw()
            finder.walk([user_dir, False, False, b""])
            finder.first_100(step=10)
        
        elif arg == 2:
            self.top.withdraw()
            finder.walk([user_dir, True, False, b""])
            finder.first_100(step=10)
        
        elif arg == 3:
            self.top.withdraw()
            finder.walk([user_dir, True, True, b""])
            finder.first_100(step=10)
        
        elif arg == 4:
            finder.step_size = 1
            finder.search_by_content = True
            self.get_input(finder, [user_dir, True, False, self.text_input])
            
        elif arg == 5:
            self.get_input(finder, [user_dir, True, False, self.text_input])
            
        elif arg == 6:
            pass
        

    def add_title(self, top, text, colour="black", font_family="Arial", size=32):
        title1 = tkinter.Label(top, fg="black", anchor="n")
        title1.config(font=(font_family, size), text=text)
        title1.grid(row=self.row_counter, sticky=tkinter.W)
        self.row_counter += 1

    def add_button(self, top, text, callback, width=30, arg=[]):
        b_size_1 = tkinter.Button(top, text=text, command=lambda: callback(arg))
        b_size_1.config(width=width)
        b_size_1.grid(row=self.row_counter, sticky=tkinter.W)
        self.row_counter += 1

    def menu(self):
        self.top = tkinter.Tk()
        self.top.title("Datadase Scripts")

        #File/folder sizes
        self.add_title(self.top, "Largest files and folders", font_family="Arial", size=14)
        self.add_button(self.top, "Search by Files", self.button_callback, arg=1)
        self.add_button(self.top, "Search by Folders", self.button_callback, arg=2)
        self.add_button(self.top, "Folders and sub folders (Slow)", self.button_callback, arg=3)
        self.add_title(self.top, " ")

        #Search for file by query
        self.add_title(self.top, "File Search", font_family="Arial", size=14)
        self.add_button(self.top, "Search files by content (Slow)", self.button_callback, arg=4)
        self.add_button(self.top, "Search files by name (fast)", self.button_callback, arg=5)
        self.add_title(self.top, " ")

        #Manage files
        self.add_title(self.top, "Manage Files", font_family="Arial", size=14)
        self.add_button(self.top, "Remove duplicate files by hash", self.button_callback, arg=6)
        self.add_title(self.top, " ")

menu = GUI()
menu.menu()

File: file-search-C++/find_largest.cpp
#include <iostream>
#include <stdio.h>
#include <string>
#include <vector>
#include <cstring>
#include <algorithm>
#include <windows.h>
#include <string>
#include <fstream>
#include <regex>
#include <map>
#include <cmath>

using namespace std;

int file_count = 0;
int dir_count = 0;
bool running = true;
string current_dir = "";
string line = "";
int filename_length;
string line_copy = "";
string filesize;
string filename;
vector<string> output_dirs = {};
vector<string> search_results = {};
vector<string> systemlinks = {};
string systemlink_path;
string path;
string extension;
vector<string> file_types = {
    ".ods",".mhtml",".ml",".otx",".xlam",".ptm",".xlsb",".ps",".xps",".xlw",
    ".xlsm",".dot",".mht",".dif",".xlt",".doc",".xltm",".txt",".dp",".psm",".csv",".odt",
    ".docx",".xml",".pam",".htm",".dotm",".xls",".xlsx",".ptx",".docm",".rtf",".pdf",".otm",
    ".htm,.html",".xltx",".prn",".wps",".df",".psx",".ps",".html",".slk",".xla",".dotx",
    ".tf",".ot",".dbf",".py"
};
vector<string> largest_folders;
string file_size = "";

map<long long, string> map_filenames;
vector<long long> largest_files;
long long int_file_size;

DWORD WINAPI output(LPVOID lpParameter) {
    int last_count;
    int fps;
    while (running == true) {
        last_count = file_count;
        system("timeout 1 > nul");
        fps = file_count - last_count;
        if (running == true) {
            cout << "Processed " << file_count << " files! FPS=" << fps << endl;
        }
    }
    return 0;
}

void search_file(string filename, string term) {
    // checks if a file contains a search term

    if (filename.find('.') != string::npos) {
        // check if extension in allowed file types
        extension = filename.substr(filename.find('.'));
        if (std::find(file_types.begin(), file_types.end(), extension) != file_types.end()) {
            // read file
            ifstream file(filename);
            int line_num = 0;
            while (getline(file, line)) {
                if (line.find(term) != std::string::npos) {
                    search_results.push_back(filename + " at line " + std::to_string(line_num));
                    return;
                }
                line_num++;
            }
        }
    }
}

string get_file_path(string search_term) {
    // get filename
    if (line.length() > 0) {
        line_copy = line;                               // copy string
        reverse(line_copy.begin(), line_copy.end());    // reverse copy
        filename_length = line_copy.find(' ');          // length of filename
        filename = line.substr(line.length() - filename_length);

        if (filename.find('.') != string::npos) {
            path = current_dir.substr(0,current_dir.length()-1)+"\\"+filename.substr(0,filename.length()-1);
            return path;
        }
    }
    return "";
}

vector<string> output_array;
int lowest;
int array_index;
int lowest_value;
int x_value;
int lowest_index;
int x_index;
string temp;

int is_number(string str) {
    if (str.length() == 0) {return false;}
    for (char const &c : str) {
        if (isdigit(c) == 0) {
            return false;
        }
    }
    return true;
}

vector<string> size_labels = {" Bytes", "KB", "MB", "GB", "TB", "PB"};
long double current_number;
int format_counter;
string format_size(long long number) {
    format_counter = 0;
    current_number = number;
    while (current_number > 1024) {
        current_number = current_number / 1024;
        format_counter++;
    }
    current_number = (long long)(current_number * 100) / 100;
    return to_string(current_number) + size_labels.at(format_counter);
}

int spaces_counter;
void get_file_data(string line, string current_dir, string search_term, bool files_only, bool filename_search, bool content_search, 
    bool maxsize, bool systemlink_search) {

    if (line.length() > 15) {
        // check if line contains file
        if (isdigit(line.at(0)) != 0 && isdigit(line.at(15)) != 0 && line.at(2) == '/') {
            if (line.find('<') == string::npos) {

                // get name and size
                spaces_counter = 0;
                for (int i = 18; i < line.length(); i++) {
                    if (line.at(i) != ' ') {
                        break;
                    }
                    spaces_counter++;
                }

                line = line.substr(18 + spaces_counter, line.length());
                filesize = line.substr(0, line.find(' '));
                filename = line.substr(filesize.length() + 1);
                filesize.erase(remove(filesize.begin(), filesize.end(), ','), filesize.end());

                // update vectors
                path = current_dir.substr(0,current_dir.length()-1)+"\\"+filename.substr(0,filename.length()-1);
                int_file_size = stoll(filesize);
                map_filenames.insert(pair<long long, string>(int_file_size, path));

                // filesize search
                if (maxsize == true) {
                    largest_files.push_back(int_file_size);
                }
                
                // filename search
                if (filename_search == true) {
                    if (line.find(search_term) != string::npos) {
                        output_dirs.push_back(path);
                    }
                }

                // content search
                if (content_search == true) {
                    search_file(path, search_term);
                }
                file_count++;

            // search for <SYMLINKD>
            } else if (systemlink_search == true && line.find("<SYMLINKD>") != string::npos) {
                //cout << "line: " << line << endl;
                systemlink_path = line.substr(line.find('['), line.length());
                systemlinks.push_back("<SYMLINKD> " + systemlink_path + "Located in: " + current_dir);
            }

        }
    }

}

void search_for_file(const char* command, string search_term, bool files_only, bool filename_search, bool content_search, bool maxsize,
        bool systemlink_search) {

    CreateThread(NULL, 0, output, NULL, 0, NULL);
    char buffer[256];
    vector<string> file_hashes = {};
    vector<string> file_names = {};
    FILE* pipe = _popen(command, "r");

    if (files_only == true) {
        search_term = ' ' + search_term;
    }

    while (fgets(buffer, sizeof buffer, pipe) != NULL) {
        try {
            // convert char to string
            line = string(buffer);

            // check if line contains a directory
            if (line.find(" Directory of") != string::npos) {
                current_dir = line.substr(14);
                dir_count++;

            } else {
                get_file_data(line, current_dir, search_term, files_only, filename_search, content_search, maxsize, systemlink_search);
            }

        } catch (int error) {
            cout << "error thrown!" << endl;
        }
    }

    // filename result
    if (filename_search == true) {
        running = false;
        cout << "\n-------- Search results --------\n" << endl;
        if (output_dirs.size() > 0) {
            for (int i = 0; i < output_dirs.size(); i++) {
                cout << output_dirs.at(i) << endl;
            }
        } else {
            cout << "No search results found!" << endl;
        }
    }

    // file content search
    if (content_search == true) {
        running = false;
        cout << "\n-------- File content search results --------\n" << endl;
        if (search_results.size() > 0) {
            for (int i = 0; i < search_results.size(); i++) {
                cout << search_results.at(i) << endl;

            }
        } else {
            cout << "No search results found!" << endl;
        }
    }

    // largest files
    if (maxsize == true) {
        running = false;
        cout << "\n-------- Largest files --------\n" << endl;

        sort(largest_files.begin(), largest_files.end());
        reverse(largest_files.begin(), largest_files.end());
        long long current;

        for (int i = 0; i < largest_files.size(); i++) {
            if (i > 200) {break;}
            current = largest_files.at(i);
            cout << format_size(current) << "|\t" << map_filenames.at(current) << endl;

        }
    }

    // systemlink search
    if (systemlink_search == true) {
        running = false;
        cout << "\n-------- Systemlink results" << endl;
        if (systemlinks.size() > 0) {
            for (int i = 0; i < systemlinks.size(); i++) {
                cout << systemlinks.at(i) << endl;
            }
        } else {
            cout << "No search results found!" << endl;
        }
    }
    
    // close handle
    _pclose(pipe);
    return;

}

void search_by_filename() {
    // menu
    cout << "V2\nMenu:\n1. Search by filename\n2. Search by file content\n3. Search by both filename and content";
    cout << "\n4. Find Largest files\n5. Find System Links\n> ";
    
    string choice;
    getline(cin, choice);

    // get user input path
    cout << "Enter path: ";
    string user_input = "";
    getline(cin, user_input);
    string command = "dir \"" + user_input + "\" /s /a";

    // search term
    string search_term = "";
    if (choice != "4" && choice != "5") {
        cout << "Enter Search term: ";
        getline(cin, search_term);
    }
    
    // convert string to char
    char buffer[270];
    strcpy(buffer, command.c_str());

    // execute command
    if (choice == "1") {
        search_for_file(buffer, search_term, false, true, false, false, false);
    } else if (choice == "2") {
        search_for_file(buffer, search_term, false, false, true, false, false);
    } else if (choice == "3") {
        search_for_file(buffer, search_term, false, true, true, false, false);
    } else if (choice == "4") {
        search_for_file(buffer, search_term, false, false, false, true, false);
    } else if (choice == "5") {
        search_for_file(buffer, search_term, false, false, false, false, true);
    }
    
}

int main() {

    // create threads
    //std::thread t1(output);

    search_by_filename();

    system("pause");
}

/*
- ask user for path

*/